Add support for unpacking tuples in args annotation

cstruct opened the pull request:
Implementation and tests for supporting unpacking tuples for `*args` annotations.

I've sneaked in some typing fixes for all tests which now pass pyright:
- missing return annotations.
- Change to keyword arguments where names where incompatible.
- Removed a duplicated test.

Please let me know if you want me to do any changes.

cstruct committed:
Fix type issues in test files.

Mostly missing return annotations. All test files passes pyright with
flying colors.

cstruct committed:
Add support for unpacking tuples in args annotation

Also handles unpacking of TypeVarTuples in the correct way but is not
supported by beartype currently.

gabrieldemarmiesse commented:
Thank you for the pull request. I'll review it and merge in the next few days and do a release. For your information, I won't be maintaining overtake, I don't have the time, I prefer to focus on my other projects Python-on-whales and Mojo(which has the functionality of overtake built-in). 

Being unpaid to do open source means I have to make some choices about what I should work on in my free time. I'll put the project in read-only with a note stating that this is unsupported. The project has an MIT license, so feel free to fork it if you need :) If someone wants to make an "overtake2", that would make me happy!

cstruct commented:
I enjoyed playing around with overtake, so I'll think long and hard about if I have the time to maintain a proper fork. Either way I fully understand your decision and wish you the best of luck!

cstruct committed:
Better handling of unpacking TypedDicts in **kwargs

If mixing unpack with explicit keyword arguments you could accidentally
get a compatible bind and no args to check.

cstruct commented:
The new commit is the result of me opening the PR from the master branch of my fork. It was not my intention to push any more maintenance burden. Please let me know if you want me to remove it from this PR in the case you want to do a finally review+release.

cstruct committed:
Support mixing pos variadic with normal positional arg funcs

cstruct committed:
Support pos args with different names and types

cstruct commented:
I'll close this. Thank you for a great library, it's been a delight working with your code! If my interest stays I'll create a `overtake2` fork. I wish you the best!

gabrieldemarmiesse commented:
Many thanks for your kind words! I hope that it helped python users see the value of dynamic dispatch and overloading :)
